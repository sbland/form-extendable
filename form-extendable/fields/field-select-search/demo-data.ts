export const defaultVal = { uid: 'a', label: 'A' };

export const demoOptions = [
  {
    uid: 'a',
    label: 'a',
  },
  {
    uid: 'b',
    label: 'b',
  },
  {
    uid: 'c',
    label: 'c',
  },
];
