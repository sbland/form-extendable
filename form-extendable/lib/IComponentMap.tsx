import { EFilterType } from '@react_db_client/constants.client-types';
import { TFieldComponentPropsAll } from './IField';
import { IFile } from './IFile';
import {
  IHeadingBool,
  IHeadingButton,
  IHeadingCustomType,
  IHeadingDate,
  IHeadingFile,
  IHeadingNumber,
  IHeadingOther,
  IHeadingReference,
  IHeadingReferenceMulti,
  IHeadingSelect,
  IHeadingSelectMulti,
  IHeadingSelectSearch,
  IHeadingTextArea,
  THeading,
  TMultiSelectValue,
} from './IHeading';

export type TComponentMapComponent<
  V,
  H extends THeading<V>
> = TFieldComponentPropsAll<V, H>;

// export type TComponentMap = Record<
//   EFilterType & string,
//   TComponentMapComponent<any>
// >;

// export type TComponentMapCustom<K extends string | number | symbol> = Record<
//   K,
//   TComponentMapComponent<never>
// >;

// export type TComponentMapFunc = <K extends never>() =>
//   | TComponentMap
//   | TComponentMapCustom<K>;

export type TComponentMap =
  | {
      [EFilterType.uid]: () => TComponentMapComponent<
        string,
        IHeadingOther<string>
      >;
      [EFilterType.text]: () => TComponentMapComponent<
        string,
        IHeadingOther<string>
      >;
      [EFilterType.embedded]: () => TComponentMapComponent<
        null,
        IHeadingOther<null>
      >;
      [EFilterType.button]: () => TComponentMapComponent<
        null,
        IHeadingButton<null>
      >;
      [EFilterType.dict]: () => TComponentMapComponent<
        Object,
        IHeadingOther<Object>
      >;
      [EFilterType.video]: () => TComponentMapComponent<
        string,
        IHeadingOther<string>
      >;
      [EFilterType.textLong]: () => TComponentMapComponent<
        string,
        IHeadingTextArea<string>
      >;
      // TODO: Select search multi
      [EFilterType.selectSearch]: () => TComponentMapComponent<
        any,
        IHeadingSelectSearch<any> | IHeadingSelectMulti<any>
      >;
      // TODO: Select search multi
      [EFilterType.reference]: () => TComponentMapComponent<
        any,
        IHeadingReference<any> | IHeadingReferenceMulti<any>
      >;
      [EFilterType.select]: () => TComponentMapComponent<
        string,
        IHeadingSelect<string>
      >;
      [EFilterType.selectMulti]: () => TComponentMapComponent<
        TMultiSelectValue,
        IHeadingSelectMulti<TMultiSelectValue>
      >;
      [EFilterType.file]: () => TComponentMapComponent<
        IFile,
        IHeadingFile<IFile>
      >;
      [EFilterType.fileMultiple]: () => TComponentMapComponent<
        IFile[],
        IHeadingFile<IFile[]>
      >;
      [EFilterType.image]: () => TComponentMapComponent<
        IFile,
        IHeadingFile<IFile>
      >;
      [EFilterType.number]: () => TComponentMapComponent<
        number,
        IHeadingNumber<number>
      >;
      [EFilterType.date]: () => TComponentMapComponent<
        Date | number | string,
        IHeadingDate<Date | number | string>
      >;
      [EFilterType.bool]: () => TComponentMapComponent<
        boolean,
        IHeadingBool<boolean>
      >;
      [EFilterType.toggle]: () => TComponentMapComponent<
        boolean,
        IHeadingBool<boolean>
      >;
    };
