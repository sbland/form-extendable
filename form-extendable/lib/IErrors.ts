import { Uid } from "@react_db_client/constants.client-types";

export enum EFormValidationError {
  REQUIRED = 'REQUIRED',
  MIN_LENGTH = 'MIN_LENGTH',
  MAX_LENGTH = 'MAX_LENGTH',
  MIN = 'MIN',
  MAX = 'MAX',
  PATTERN = 'PATTERN',
  EMAIL = 'EMAIL',
  MATCH = 'MATCH',
  CUSTOM = 'CUSTOM',
  UNKNOWN = 'UNKNOWN',
}

export interface IFormFieldValidationError {
  type: EFormValidationError;
  message: string;
  field?: Uid;
  value?: any;
}
