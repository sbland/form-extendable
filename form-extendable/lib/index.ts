export * from './IField';
export * from './IHeading';
export * from './IComponentMap';
export * from './IFormData';
export * from './IOpt';
export * from './IObj';
export * from './IPopupProps';
export { EFormValidationError, IFormFieldValidationError } from './IErrors';
