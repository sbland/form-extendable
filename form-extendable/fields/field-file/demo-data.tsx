import React from 'react';
import { IPopupProps } from '@form-extendable/lib';
import {
  EFileType,
  EFilterType,
  IFile,
} from '@react_db_client/constants.client-types';
import { IFieldFileProps } from './field-file';
import { IFieldFileMultipleProps } from './field-file-multiple';

export const DEMO_IMAGE_FILES_DATA: IFile[] = [
  {
    uid: 'a',
    filePath: '',
    label: 'FileA',
    name: 'bit-logo.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'b',
    filePath: 'extensions-icons',
    label: 'FileB',
    name: 'vue_grey.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'c',
    filePath: 'Community/app-components',
    label: 'FileC',
    name: 'card.jpg',
    fileType: EFileType.IMAGE,
  },
];

export const DEMO_FILES_DATA: IFile[] = [
  {
    uid: 'a',
    filePath: '',
    label: 'FileA with a really long name',
    name: 'bit-logo.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'b',
    filePath: 'extensions-icons',
    label: 'FileB',
    name: 'vue_grey.svg',
    fileType: EFileType.IMAGE,
  },
  {
    uid: 'c',
    filePath: 'Community/app-components',
    label: 'FileC',
    name: 'card.jpg',
    fileType: EFileType.IMAGE,
  },
];

export const DEMO_FILES_DATA_MANY = [
  ...DEMO_FILES_DATA,
  ...Array(100)
    .fill(0)
    .map((_, i) => ({
      uid: `file_${i}`,
      filePath: '',
      label: `FileA with a really long name ${i}`,
      name: 'bit-logo.svg',
      fileType: EFileType.IMAGE,
    })),
];

export const DEMO_IMAGE_FILES_MANY = [
  ...DEMO_IMAGE_FILES_DATA,
  ...Array(100)
    .fill(0)
    .map((_, i) => ({
      uid: `file_${i}`,
      filePath: '',
      label: `FileA with a really long name ${i}`,
      name: `bit-logo_${i}.svg`,
      fileType: EFileType.IMAGE,
    })),
];

export const PopupPanel: React.FC<IPopupProps> = ({
  children,
  isOpen,
  handleClose,
}) => (isOpen ? <>{children}</> : <></>);

const onChange: IFieldFileProps['onChange'] = () => {};
const onChangeMulti: IFieldFileMultipleProps['onChange'] = () => {};

export const dummyProps: IFieldFileProps = {
  uid: 'uid',
  label: 'File Field',
  type: EFilterType.file,
  multiple: false,
  onChange,
  fileType: EFileType.IMAGE,
  value: DEMO_IMAGE_FILES_DATA[0],
  fileServerUrl: 'https://static.bit.dev',
  asyncGetFiles: (metaData) => async () => DEMO_IMAGE_FILES_DATA,
  PopupPanel,
  asyncFileUpload: () => async () => {},
};

export const dummyPropsImagesMany: IFieldFileMultipleProps = {
  ...dummyProps,
  onChange: onChangeMulti,
  type: EFilterType.fileMultiple,
  defaultValue: [],
  multiple: true,
  value: [...DEMO_IMAGE_FILES_MANY].slice(0, 50),
  asyncGetFiles: (metaData) => async () =>
    [...DEMO_IMAGE_FILES_MANY].slice(50, 60),
};

export const dummyPropsDocs: IFieldFileProps = {
  ...dummyProps,
  fileType: EFileType.DOCUMENT,
  value: DEMO_FILES_DATA[0],
  asyncGetFiles: (metaData) => async () => DEMO_FILES_DATA,
};

export const dummyPropsDocsMany: IFieldFileMultipleProps = {
  ...dummyPropsDocs,
  onChange: onChangeMulti,
  type: EFilterType.fileMultiple,
  defaultValue: [],
  multiple: true,
  value: [...DEMO_FILES_DATA_MANY].slice(0, 50),
  asyncGetFiles: (metaData) => async () =>
    [...DEMO_FILES_DATA_MANY].slice(50, 60),
};
