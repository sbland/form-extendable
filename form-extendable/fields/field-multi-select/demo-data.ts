import { IHeadingSelectMulti } from '@form-extendable/lib';
import { EFilterType } from '@react_db_client/constants.client-types';
import { TFieldMultiSelect } from './field-multi-select';

export const demoOptions = [
  {
    uid: 'a',
    label: 'A',
  },
  {
    uid: 'b',
    label: 'B',
  },
  {
    uid: 'c',
    label: 'C',
  },
];

export const demoHeading: IHeadingSelectMulti = {
  uid: 'multiSelectListShowAll',
  label: 'Multi Select Show All',
  type: EFilterType.selectMulti,
  selectType: 'showall',
  options: demoOptions,
  required: true,
};

export const props: TFieldMultiSelect = {
  type: EFilterType.selectMulti,
  label: demoHeading.label,
  uid: demoHeading.uid,
  unit: 'unit',
  onChange: () => {},
  options: demoOptions,
  value: [demoOptions[0].uid],
  required: demoHeading.required,
};
