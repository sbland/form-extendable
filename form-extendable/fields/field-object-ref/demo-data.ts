import { IObj } from '@form-extendable/lib';

export const defaultVal = '123';

export const demoResults: IObj[] = [
  {
    _id: '123',
    uid: 'foo',
    label: 'Foo',
  },
  {
    _id: '456',
    uid: 'bar',
    label: 'Bar',
  },
];
