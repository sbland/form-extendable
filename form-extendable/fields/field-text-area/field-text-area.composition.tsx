import React, { useState } from 'react';
import { CompositionWrapDefault } from '@react_db_client/helpers.composition-wraps';
import { WrapFieldComponent } from '@form-extendable/composition-helpers';
import { EFilterType } from '@react_db_client/constants.client-types';
import { FieldTextArea, TFieldTextAreaProps } from './field-text-area';
import { defaultVal } from './demo-data';

const onChange = () => {};

const defaultProps: TFieldTextAreaProps = {
  type: EFilterType.textLong,
  label: 'Text Area Field',
  uid: 'uid',
  unit: 'unit',
  value: defaultVal,
  onChange,
};

export const BasicFieldTextArea = () => (
  <CompositionWrapDefault height="4rem" width="16rem">
    <WrapFieldComponent>
      <FieldTextArea {...defaultProps} required />
    </WrapFieldComponent>
  </CompositionWrapDefault>
);

export const WithStyleOverride = () => (
  <CompositionWrapDefault height="4rem" width="16rem">
    <WrapFieldComponent>
      <FieldTextArea
        {...defaultProps}
        styleOverrides={{ color: 'red', width: '100%' }}
      />
    </WrapFieldComponent>
  </CompositionWrapDefault>
);

export const FieldTextAreaMultiple = () => (
  <CompositionWrapDefault height="4rem" width="16rem">
    <WrapFieldComponent>
      <FieldTextArea {...defaultProps} required />
      <FieldTextArea {...defaultProps} required />
      <FieldTextArea {...defaultProps} required />
    </WrapFieldComponent>
  </CompositionWrapDefault>
);

export const FieldTextAreaMultipleHoriz = () => (
    <CompositionWrapDefault height="4rem" width="16rem">
      <div style={{ display: 'flex' }}>
        <WrapFieldComponent>
          <FieldTextArea {...defaultProps} required />
          <FieldTextArea {...defaultProps} required />
          <FieldTextArea {...defaultProps} required />
        </WrapFieldComponent>
      </div>
    </CompositionWrapDefault>
  );
