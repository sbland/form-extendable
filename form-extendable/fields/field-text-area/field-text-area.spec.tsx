import React from 'react';

import { render } from '@testing-library/react';
import { IFieldComponentProps, IHeadingTextArea } from '@form-extendable/lib';
import { EFilterType } from '@react_db_client/constants.client-types';
import { FieldTextArea } from './field-text-area';
import * as compositions from './field-text-area.composition';
import { defaultVal } from './demo-data';

const onChange = jest.fn();

const defaultProps: IFieldComponentProps<string, IHeadingTextArea> = {
  uid: 'uid',
  unit: 'unit',
  value: defaultVal,
  onChange,
  type: EFilterType.textLong,
  label: 'Text Area',
};

describe('field-text-area', () => {
  beforeEach(() => {
    onChange.mockClear();
  });
  test('Renders', () => {
    render(<FieldTextArea {...defaultProps} />);
  });

  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        render(<Composition />);
      });
    });
  });
  // describe('Unit Tests', () => {
  //   let component;
  //   let spyRef;
  //   beforeEach(() => {
  //     spyRef = jest.spyOn(React, 'useRef');
  //     spyRef.mockReturnValue({ current: { scrollHeight: 100 } });
  //     component = shallow(<FieldTextArea {...defaultProps} />);
  //   });
  //   describe('Auto size', () => {
  //     test('should autosize to initial content', async () => {
  //       spyRef.mockReturnValue({ current: { scrollHeight: 100 } });
  //       component.update();
  //       const textArea = component.find('textarea');
  //       expect(textArea.props().style.height).toEqual('auto');
  //       // await new Promise((resolve) => setTimeout(resolve));
  //       // TODO: THis works but not in test.
  //       // expect(textArea.props().style.height).toEqual(10);
  //     });
  //   });
  // });
});
