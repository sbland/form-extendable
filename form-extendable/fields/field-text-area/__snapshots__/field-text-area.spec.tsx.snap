// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field-text-area shallow renders Matches Snapshot 1`] = `undefined`;
